### PR TITLE
Consensus simulation:

### PR DESCRIPTION
**Pull request by vinniefalco**
* Define Peer concept in BasicNetwork
* Add incremental consensus simulations

### CODE REVIEW

File: src/ripple/test/BasicNetwork.h [hunk 1]
```diff
@@ -197,10 +199,10 @@ class BasicNetwork
             boost::intrusive::make_multiset<msg,
                 boost::intrusive::constant_time_size<false>>::type;
 
-        std::unordered_map<Peer*, by_to_list> by_to_;
-        std::unordered_map<Peer*, by_from_list> by_from_;
-        by_when_set by_when_;
         qalloc alloc_;
+        by_when_set by_when_;
+        std::unordered_map<Peer, by_to_list> by_to_;
+        std::unordered_map<Peer, by_from_list> by_from_;
```
Comment (seelabs) [inline src/ripple/test/BasicNetwork.h]: We're storing copies of peer multiple times. I get that `Peer` is often a pointer type, in which case this is fine. Are we still fine with this when `Peer` is a more expensive type?

**Comment by ximinez** [comment]
Assuming CI passes, this looks good. :+1:

File: src/ripple/test/BasicNetwork.h (hunk 1 above)
Comment (vinniefalco) [inline src/ripple/test/BasicNetwork.h]: An expensive type is a poor choice for `Peer`.

Comment (seelabs) [inline src/ripple/test/BasicNetwork.h]: Maybe add that to the doc specifying peer requirements. Peer types should be cheap, and point out that a pointer meets the requirements.

**Comment by seelabs** [comment]
:+1: Quicker review than usual, but it all looks low risk and the code looks good.

**Comment by ximinez** [comment]
Merged 49c86768e617c071433d4455d371971bb76ed297

**Pull request closed without merging**